pkg/tcp/proxy.go

```diff
@@ -10,28 +10,29 @@ import (
 
 // Proxy forwards a TCP request to a TCP service.
 type Proxy struct {
-	target           *net.TCPAddr
+	address          string
 	terminationDelay time.Duration
 }
 
 // NewProxy creates a new Proxy.
 func NewProxy(address string, terminationDelay time.Duration) (*Proxy, error) {
-	tcpAddr, err := net.ResolveTCPAddr("tcp", address)
-	if err != nil {
-		return nil, err
-	}
-
-	return &Proxy{target: tcpAddr, terminationDelay: terminationDelay}, nil
+	return &Proxy{address: address, terminationDelay: terminationDelay}, nil
 }
 
 // ServeTCP forwards the connection to a service.
 func (p *Proxy) ServeTCP(conn WriteCloser) {
 	log.Debugf("Handling connection from %s", conn.RemoteAddr())
 
+	tcpAddr, err := net.ResolveTCPAddr("tcp", p.address)
```

Comment:
Could we store the` target   *net.TCPAddr` in the `Proxy` struct, with a timeout time? if the timeout is reached, then re-resolve the TCP address?

That way we could "cache" the address for a short-ish period of time, reduce DNS lookups, and would probably reduce the overhead of multiple resolutions.

I was thinking maybe a 5 second timeout? just something that if 1000 requests come in at once, we still only resolve once, but if 1000 come in over the course of a minute, we are still updating the address 20 times, so it will respond to dynamic changes fairly quickly.  We may want this to be configurable, idk.

Thoughts?

Reply:
Yes, this could work well If we make it configurable. We could even allow a zero value for "never update" that would be like keeping the current behavior.

Reply:
@dtomcej I added this but instead of never updating I only left an always updating alternative with a negative value, I think it makes more sense.

If this is reasonable I'll add the missing docs, examples and everything else

Reply:
Looks good to me.